ch_pipeline/analysis/flagging.py

```diff
@@ -1948,11 +1949,57 @@ def process(self, raw_mask):
             exp_mask = raw_mask
         else:
             exp_mask = dcontainers.empty_like(raw_mask)
+            # Make sure that this mask is all zero
+            exp_mask.mask[:] = 0
 
-        exp_mask.mask[:] = mexp
+        exp_mask.mask[:] |= mexp
+
+        # Log the new percent of data masked
+        og_frac = 100.0 * original_mask_sum / mask.size
+        drop_frac = 100.0 * exp_mask.mask[:].sum() / mask.size
+        self.log.info(f"Originally flagged {og_frac:.2f}% of data.")
+        self.log.info(f"Flagged {drop_frac:.2f}% of data after expansion.")
 
         return exp_mask
 
+    def _padded_mask_hook(self, mask: np.ndarray) -> np.ndarray:
+        """Overwrite to modify creation of the padded mask."""
+        nfreq, ntime = mask.shape
+        mraw = np.zeros((nfreq, ntime + 2 * self.nexpand), dtype=bool)
+        mraw[:, self.nexpand : -self.nexpand] = mask
+
+        return mraw
+
+
+class ExpandGainMask(ExpandMask):
+    """Expand a Bad Gain Mask to account for gain interpolation.
+
+    Since bad calibration will only happen once per day, only expand
+    the gains in time for regions where they are consistently bad from
+    the start or until the end of the day.
+    """
+
+    def _padded_mask_hook(self, mask: np.ndarray) -> np.ndarray:
+        """Create the padded mask to expand."""
+        # Initialize the empty mask. We don't want to combine with the
+        # original mask until after expanding
+        nfreq, ntime = mask.shape
+        mraw = np.zeros((nfreq, ntime + 2 * self.nexpand), dtype=bool)
+
+        # Indices of the first and last unmasked samples along the time axis.
+        time_start_idx = mask.argmin(axis=1) + self.nexpand
+        time_end_idx = mraw.shape[1] - mask[:, ::-1].argmin(axis=1) - self.nexpand
+
+        # Only take frequencies where there is masking in the first or last samples
+        freq_start_idx = time_start_idx > self.nexpand
+        freq_end_idx = time_end_idx < mraw.shape[1] - self.nexpand
```

Comment:
I find these variable names confusing, since `freq_start_idx` and `freq_end_idx` are actually boolean masks, while `time_start_idx` and `time_end_idx` are arrays of indices.